File: server/src/server/user_game_instance/game_core/scenes_core/scene_engine.py
from __future__ import annotations
from pydantic import BaseModel
from server.user_game_instance.theme_0.scenes.intro_scene.intro_scene import IntroScene
from server.user_game_instance.theme_0.scenes.main_menu_scene.main_menu_scene import MainMenuScene 
from server.user_game_instance.theme_0.scenes.login_scene.login_scene import LoginScene 
from server.user_game_instance.theme_0.scenes.world_generation_scene.world_generation_scene import WorldGenerationScene 
from server.user_game_instance.theme_0.scenes.main_scene.main_scene import MainScene 
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from server.user_game_instance.user_game_instance import UserGameInstance

class SceneEngine(BaseModel):

    scenes : dict = {}
    current_scene : str = ""

    def __init__(self) -> None:
        super().__init__()
        
        self.scenes["IntroScene"] = IntroScene()
        self.scenes["MainMenuScene"] = MainMenuScene()
        self.scenes["LoginScene"] = LoginScene()
        self.scenes["WorldGenerationScene"] = WorldGenerationScene()
        self.scenes["MainScene"] = MainScene()

        self.current_scene = "IntroScene"
   
    def update_current_scene(self, user_game_instance : UserGameInstance) -> None:
        if self.current_scene in self.scenes:
            self.scenes[self.current_scene].update(user_game_instance)

    def render_current_scene(self, user_game_instance: UserGameInstance) -> None:
        if self.current_scene in self.scenes:
            self.scenes[self.current_scene].render(user_game_instance)

File: client/src/game/core/scenes_core/scene_engine.cpp
/*
 * Copyright 2024 Andreas Åkerberg.
 */
#include "scene_engine.hpp"
#include "game/core/engine.hpp"
#include "theme_0/scenes/intro_scene/intro_scene.hpp"
#include "theme_0/scenes/main_menu_scene/main_menu_scene.hpp"
#include "theme_0/scenes/login_scene/login_scene.hpp"
#include "theme_0/scenes/world_generation_scene/world_generation_scene.hpp"
#include "theme_0/scenes/main_scene/main_scene.hpp"

namespace JoD {
void SceneEngine::AddScene(std::string_view sceneName, IScene& scene) {
  m_scenes.insert({Hash(sceneName), scene});
}

void SceneEngine::InitializeScenes() {
  AddScene("IntroScene", _<IntroScene>());
  AddScene("MainMenuScene", _<MainMenuScene>());
  AddScene("LoginScene", _<LoginScene>());
  AddScene(
    "WorldGenerationScene",
    _<WorldGenerationScene>());
  AddScene("MainScene", _<MainScene>());
}

void SceneEngine::UpdateCurrentScene() {
  if (m_scenes.contains(m_currentScene))
    m_scenes.at(m_currentScene).Update();
}

void SceneEngine::RenderCurrentScene() {
  if (m_scenes.contains(m_currentScene))
    m_scenes.at(m_currentScene).Render();
}

void SceneEngine::GoToScene(std::string_view newSceneName) {
  m_currentScene = Hash(newSceneName);
  if (m_scenes.contains(m_currentScene)) {
    _<Engine>().ResetInputState();
    m_scenes.at(m_currentScene).OnEnter();
  }
}

Size SceneEngine::GetWorldAreaSize() const {
  return _<Engine>().GetWorldAreaSize();
}

RID SceneEngine::AllocateNewImage() const {
  return _<Engine>().AllocateNewImage();
}

void SceneEngine::DrawImage(RID rid, int imageNameHash, const BoxF &area,
                            ColorF color, bool repeatTexture,
                            SizeF textureFillAmount,
                            bool noPixelEffect) const {
  _<Engine>().DrawImage(
    rid, imageNameHash, area, color, repeatTexture,
    textureFillAmount, noPixelEffect);
}

void SceneEngine::DrawImage(RID rid, std::string_view imageName,
                            const BoxF &area,
                            ColorF color, bool repeatTexture,
                            SizeF textureFillAmount,
                            bool noPixelEffect) const {
  _<Engine>().DrawImage(
    rid, imageName, area, color,repeatTexture,
    textureFillAmount, noPixelEffect);
}

RID SceneEngine::AllocateNewString() const {
  return _<Engine>().AllocateNewString();
}

void SceneEngine::DrawString(int id,
                             std::string_view text,
                             Point2F position,
                             ColorF color,
                             bool centerAlign,
                             FontSizes fontSize) const {
  _<Engine>().DrawString(
    id, text, position, color, centerAlign,
    fontSize);
}

RID SceneEngine::AllocateNewTile() const {
  return _<Engine>().AllocateNewTile();
}

bool SceneEngine::AnyMouseButtonIsPressed() const {
  return _<Engine>().AnyMouseButtonisPressed();
}

bool SceneEngine::AnyKeyIsPressed() const {
  return _<Engine>().AnyKeyIsPressed();
}

void SceneEngine::StopEngine() {
  _<Engine>().SetRunning(false);
}
}
